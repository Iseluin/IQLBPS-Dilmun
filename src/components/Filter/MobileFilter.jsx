import { Dialog, Disclosure, Transition } from '@headlessui/react';
import React, { Fragment } from 'react';
import { XMarkIcon } from '@heroicons/react/24/outline';
import { MinusIcon, PlusIcon } from '@heroicons/react/20/solid';

const MobileFilter = ({
  mobileFiltersOpen,
  setMobileFiltersOpen,
  checkFilters,
  radioFilters,
}) => {
  return (
    <Transition.Root show={mobileFiltersOpen} as={Fragment}>
      <Dialog
        as="div"
        className="relative z-40 lg:hidden"
        onClose={setMobileFiltersOpen}
      >
        <Transition.Child
          as={Fragment}
          enter="transition-opacity ease-linear duration-300"
          enterFrom="opacity-0"
          enterTo="opacity-100"
          leave="transition-opacity ease-linear duration-300"
          leaveFrom="opacity-100"
          leaveTo="opacity-0"
        >
          <div className="fixed inset-0 bg-primary bg-opacity-25" />
        </Transition.Child>

        <div className="fixed inset-0 z-40 flex">
          <Transition.Child
            as={Fragment}
            enter="transition ease-in-out duration-300 transform"
            enterFrom="translate-x-full"
            enterTo="translate-x-0"
            leave="transition ease-in-out duration-300 transform"
            leaveFrom="translate-x-0"
            leaveTo="translate-x-full"
          >
            <Dialog.Panel className="relative ml-auto flex h-full w-full max-w-xs flex-col overflow-y-auto bg-background py-4 pb-12 shadow-xl">
              <div className="flex items-center justify-between px-4">
                <h2 className="text-lg font-medium text-primary">Filters</h2>
                <button
                  type="button"
                  className="-mr-2 flex h-10 w-10 items-center justify-center rounded-md bg-background p-2 text-primary"
                  onClick={() => setMobileFiltersOpen(false)}
                >
                  <span className="sr-only">Close menu</span>
                  <XMarkIcon className="h-6 w-6" aria-hidden="true" />
                </button>
              </div>

              {/* Filters */}
              <form className="mt-4 border-t border-gray-200">
                {checkFilters.map((section) => (
                  <Disclosure
                    as="div"
                    key={section.id}
                    className="border-t border-gray-200 px-4 py-6"
                    defaultOpen
                  >
                    {({ open }) => (
                      <>
                        <h3 className="-mx-2 -my-3 flow-root">
                          <Disclosure.Button className="flex w-full items-center justify-between bg-background px-2 py-3 text-primary hover:text-secondary">
                            <span className="font-medium text-primary">
                              {section.name}
                            </span>
                            <span className="ml-6 flex items-center">
                              {open ? (
                                <MinusIcon
                                  className="h-5 w-5"
                                  aria-hidden="true"
                                />
                              ) : (
                                <PlusIcon
                                  className="h-5 w-5"
                                  aria-hidden="true"
                                />
                              )}
                            </span>
                          </Disclosure.Button>
                        </h3>
                        <Disclosure.Panel className="pt-6">
                          <div className="space-y-6">
                            {section.options.map((option, optionIdx) => (
                              <div
                                key={option.value}
                                className="flex items-center"
                              >
                                <input
                                  id={`filter-mobile-${section.id}-${optionIdx}`}
                                  name={`${section.id}[]`}
                                  defaultValue={option.value}
                                  type="checkbox"
                                  defaultChecked={option.checked}
                                  className="h-4 w-4 rounded border-gray-300 text-primary focus:ring-secondary"
                                />
                                <label
                                  htmlFor={`filter-mobile-${section.id}-${optionIdx}`}
                                  className="ml-3 min-w-0 flex-1 text-primary select-none"
                                >
                                  {option.label}
                                </label>
                              </div>
                            ))}
                          </div>
                        </Disclosure.Panel>
                      </>
                    )}
                  </Disclosure>
                ))}

                {radioFilters.map((section) => (
                  <Disclosure
                    as="div"
                    key={section.id}
                    className="border-t border-gray-200 px-4 py-6"
                  >
                    {({ open }) => (
                      <>
                        <h3 className="-mx-2 -my-3 flow-root">
                          <Disclosure.Button className="flex w-full items-center justify-between bg-background px-2 py-3 text-primary hover:text-secondary">
                            <span className="font-medium text-primary">
                              {section.name}
                            </span>
                            <span className="ml-6 flex items-center">
                              {open ? (
                                <MinusIcon
                                  className="h-5 w-5"
                                  aria-hidden="true"
                                />
                              ) : (
                                <PlusIcon
                                  className="h-5 w-5"
                                  aria-hidden="true"
                                />
                              )}
                            </span>
                          </Disclosure.Button>
                        </h3>
                        <Disclosure.Panel className="pt-6">
                          <div className="space-y-6">
                            <div>
                              <h3 className="sr-only">Minimum Price</h3>
                              <input
                                type="number"
                                name="min-price"
                                id="mobile-min-price"
                                placeholder="Min"
                                className="border border-gray-300 rounded-md px-3 py-2 w-1/3 mr-1 focus:outline-primary"
                              />
                              <h3 className="sr-only">Maximum Price</h3>
                              <input
                                type="number"
                                name="max-price"
                                id="mobile-max-price"
                                placeholder="Max"
                                className="border border-gray-300 rounded-md px-3 py-2 w-1/3 focus:outline-primary"
                              />
                              <button
                                type="button"
                                className="bg-primary text-background px-3 py-1 rounded-md ml-5"
                              >
                                Apply
                              </button>
                            </div>
                            {section.options.map((option, optionIdx) => (
                              <div
                                key={option.value.min}
                                className="flex items-center"
                              >
                                <input
                                  id={`filter-mobile-${section.id}-${optionIdx}`}
                                  name={`${section.id}[]`}
                                  defaultValue={option.value}
                                  type="radio"
                                  defaultChecked={option.checked}
                                  className="h-4 w-4 rounded border-gray-300 text-primary focus:ring-secondary"
                                />
                                <label
                                  htmlFor={`filter-mobile-${section.id}-${optionIdx}`}
                                  className="ml-3 min-w-0 flex-1 text-primary select-none"
                                >
                                  {option.label}
                                </label>
                              </div>
                            ))}
                          </div>
                        </Disclosure.Panel>
                      </>
                    )}
                  </Disclosure>
                ))}
              </form>
            </Dialog.Panel>
          </Transition.Child>
        </div>
      </Dialog>
    </Transition.Root>
  );
};

export default MobileFilter;
