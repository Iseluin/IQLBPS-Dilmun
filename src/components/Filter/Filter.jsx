import { Fragment, useState } from 'react';
import { Disclosure } from '@headlessui/react';
import { MinusIcon, PlusIcon } from '@heroicons/react/20/solid';
import { useTranslation } from 'react-i18next';
import MobileFilter from './MobileFilter';

const Filter = ({ mobileFiltersOpen, setMobileFiltersOpen, handleFilter }) => {
  const { t } = useTranslation();
  const [checkFilters, setCheckFilters] = useState([
    {
      id: 'category',
      name: t('filter.category.title'),
      options: [
        { value: 'men', label: t('filter.category.men'), checked: false },
        { value: 'women', label: t('filter.category.women'), checked: false },
        { value: 'kids', label: t('filter.category.kids'), checked: false },
        { value: 'toys', label: t('filter.category.toys'), checked: false },
        {
          value: 'electronics',
          label: t('filter.category.electronics'),
          checked: false,
        },
        { value: 'home', label: t('filter.category.home'), checked: false },
      ],
    },
    {
      id: 'type',
      name: t('filter.type.title'),
      options: [
        { value: 'new', label: t('filter.type.new'), checked: false },
        { value: 'used', label: t('filter.type.used'), checked: false },
        { value: 'donated', label: t('filter.type.donated'), checked: false },
        { value: 'crafted', label: t('filter.type.crafted'), checked: false },
      ],
    },
  ]);
  const [radioFilters, setRadioFilters] = useState([
    {
      id: 'price',
      name: t('filter.price'),
      options: [
        { value: { min: 1, max: 25 }, label: '1$ - 25$', checked: false },
        { value: { min: 26, max: 50 }, label: '26$ - 50$', checked: false },
        {
          value: { min: 51, max: 100 },
          label: '51$ - 100$',
          checked: false,
        },
        {
          value: { min: 101, max: 200 },
          label: '101$ - 200$',
          checked: false,
        },
        {
          value: { min: 201, max: 400 },
          label: '201$ - 400$',
          checked: false,
        },
        {
          value: { min: 401, max: 800 },
          label: '401$ - 800$',
          checked: false,
        },
        {
          value: { min: 800 },
          label: '+800$',
          checked: false,
        },
      ],
    },
  ]);

  const handleOnCheckChange = (sectionId, optionIndex) => {
    const updatedFilters = checkFilters.map((section) => {
      if (section.id === sectionId) {
        const updatedOptions = section.options.map((option, index) => {
          if (index === optionIndex) {
            return { ...option, checked: !option.checked };
          }
          return option;
        });
        return { ...section, options: updatedOptions };
      }
      return section;
    });
    setCheckFilters(updatedFilters);
  };

  const handleOnRadioChange = (sectionId, optionIndex) => {
    const updatedFilters = radioFilters.map((section) => {
      if (section.id === sectionId) {
        const updatedOptions = section.options.map((option, index) => {
          if (index === optionIndex) {
            return { ...option, checked: !option.checked };
          }
          return { ...option, checked: false };
        });
        return { ...section, options: updatedOptions };
      }
      return section;
    });
    setRadioFilters(updatedFilters);
  };
  const selectedCheckFilters = checkFilters.map((section) => {
    const selectedOptions = section.options.filter((option) => option.checked);
    return { ...section, options: selectedOptions };
  });
  const selectedRadioFilters = radioFilters.map((section) => {
    const selectedOptions = section.options.filter((option) => option.checked);
    return { ...section, options: selectedOptions };
  });

  const categoryOptions = selectedCheckFilters[0].options.length
    ? selectedCheckFilters[0].options.map((option) => option.label)
    : [];
  const typeOptions = selectedCheckFilters[0].options.length
    ? selectedCheckFilters[1].options.map((option) => option.label)
    : [];
  const priceOptions = selectedRadioFilters[0].options.length
    ? selectedRadioFilters[0].options[0].value
    : {};

  handleFilter(categoryOptions, typeOptions, priceOptions);
  return (
    <div data-testid="filter" className="bg-background">
      <div>
        {/* Mobile filter dialog */}
        <MobileFilter
          mobileFiltersOpen={mobileFiltersOpen}
          setMobileFiltersOpen={setMobileFiltersOpen}
          checkFilters={checkFilters}
          radioFilters={radioFilters}
        />

        {/* Filters */}
        <form className="hidden lg:block p-4">
          {checkFilters.map((section) => (
            <Disclosure
              as="div"
              key={section.id}
              className="border-b border-gray-200 py-6"
              defaultOpen={section.id === 'category'}
            >
              {({ open }) => (
                <>
                  <h3 className="-my-3 flow-root">
                    <Disclosure.Button className="flex w-full items-center justify-between bg-background py-3 text-sm text-primary hover:text-secondary">
                      <span className="font-medium text-primary">
                        {section.name}
                      </span>
                      <span className="ml-6 flex items-center">
                        {open ? (
                          <MinusIcon className="h-5 w-5" aria-hidden="true" />
                        ) : (
                          <PlusIcon className="h-5 w-5" aria-hidden="true" />
                        )}
                      </span>
                    </Disclosure.Button>
                  </h3>
                  <Disclosure.Panel className="pt-6">
                    <div className="space-y-4">
                      {section.options.map((option, optionIdx) => (
                        <div key={option.value} className="flex items-center">
                          <input
                            id={`filter-${section.id}-${optionIdx}`}
                            name={`${section.id}[]`}
                            defaultValue={option.value}
                            type="checkbox"
                            defaultChecked={option.checked}
                            className="h-4 w-4 rounded border-gray-300 text-primary focus:ring-secondary"
                            onChange={() =>
                              handleOnCheckChange(section.id, optionIdx)
                            }
                          />
                          <label
                            htmlFor={`filter-${section.id}-${optionIdx}`}
                            className="ml-3 text-sm text-primary select-none"
                          >
                            {option.label}
                          </label>
                        </div>
                      ))}
                    </div>
                  </Disclosure.Panel>
                </>
              )}
            </Disclosure>
          ))}
          {radioFilters.map((section) => (
            <Disclosure
              as="div"
              key={section.id}
              className="border-b border-gray-200 py-6"
            >
              {({ open }) => (
                <>
                  <h3 className="-my-3 flow-root">
                    <Disclosure.Button className="flex w-full items-center justify-between bg-background py-3 text-primary hover:text-secondary">
                      <span className="font-medium text-primary">
                        {section.name}
                      </span>
                      <span className="ml-6 flex items-center">
                        {open ? (
                          <MinusIcon className="h-5 w-5" aria-hidden="true" />
                        ) : (
                          <PlusIcon className="h-5 w-5" aria-hidden="true" />
                        )}
                      </span>
                    </Disclosure.Button>
                  </h3>
                  <Disclosure.Panel className="pt-6">
                    <div className="space-y-6">
                      <div>
                        <h3 className="sr-only">Minimum Price</h3>
                        <input
                          type="number"
                          name="min-price"
                          id="min-price"
                          placeholder="Min"
                          className="border border-gray-300 rounded-md px-3 py-2 w-1/3 md:w-1/4 mr-1 focus:outline-primary"
                        />
                        <h3 className="sr-only">Maximum Price</h3>
                        <input
                          type="number"
                          name="max-price"
                          id="max-price"
                          placeholder="Max"
                          className="border border-gray-300 rounded-md px-3 py-2 w-1/3 md:w-1/4 focus:outline-primary"
                        />
                        <button
                          type="button"
                          className="bg-primary text-background px-3 py-1 rounded-md ml-5"
                        >
                          {t('filter.apply')}
                        </button>
                      </div>
                      {section.options.map((option, optionIdx) => (
                        <div
                          key={option.value.min}
                          className="flex items-center"
                        >
                          <input
                            id={`filter-${section.id}-${optionIdx}`}
                            name={`${section.id}[]`}
                            defaultValue={option.value}
                            type="radio"
                            defaultChecked={option.checked}
                            className="h-4 w-4 rounded border-gray-300 text-primary focus:ring-secondary"
                            onChange={() =>
                              handleOnRadioChange(section.id, optionIdx)
                            }
                          />
                          <label
                            htmlFor={`filter-${section.id}-${optionIdx}`}
                            className="ml-3 min-w-0 flex-1 text-primary select-none"
                          >
                            {option.label}
                          </label>
                        </div>
                      ))}
                    </div>
                  </Disclosure.Panel>
                </>
              )}
            </Disclosure>
          ))}
        </form>
      </div>
    </div>
  );
};

export default Filter;
