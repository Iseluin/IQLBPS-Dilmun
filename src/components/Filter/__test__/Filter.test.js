
import { render, screen } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import Filter from '../Filter';

test('Sign In Page Renders Correctly', () => {
  render(
    <Router>
      <Filter />
    </Router>
  );
  expect(screen.getByTestId('filter')).toMatchSnapshot();
});
