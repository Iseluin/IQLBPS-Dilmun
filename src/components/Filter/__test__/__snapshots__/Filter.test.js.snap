// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sign In Page Renders Correctly 1`] = `
<div
  class="bg-background"
  data-testid="filter"
>
  <div>
    <main
      class="mx-auto max-w-7xl px-4 sm:px-6 lg:px-8"
    >
      <div
        class="flex items-baseline justify-between border-b border-gray-200 pt-24 pb-6"
      >
        <h1
          class="text-4xl font-bold tracking-tight text-primary"
        >
          New Arrivals
        </h1>
        <div
          class="flex items-center"
        >
          <button
            class="-m-2 ml-5 p-2 text-secondary hover:text-primary sm:ml-7"
            type="button"
          >
            <span
              class="sr-only"
            >
              View grid
            </span>
            <svg
              aria-hidden="true"
              class="h-5 w-5"
              fill="currentColor"
              viewBox="0 0 20 20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M4.25 2A2.25 2.25 0 002 4.25v2.5A2.25 2.25 0 004.25 9h2.5A2.25 2.25 0 009 6.75v-2.5A2.25 2.25 0 006.75 2h-2.5zm0 9A2.25 2.25 0 002 13.25v2.5A2.25 2.25 0 004.25 18h2.5A2.25 2.25 0 009 15.75v-2.5A2.25 2.25 0 006.75 11h-2.5zm9-9A2.25 2.25 0 0011 4.25v2.5A2.25 2.25 0 0013.25 9h2.5A2.25 2.25 0 0018 6.75v-2.5A2.25 2.25 0 0015.75 2h-2.5zm0 9A2.25 2.25 0 0011 13.25v2.5A2.25 2.25 0 0013.25 18h2.5A2.25 2.25 0 0018 15.75v-2.5A2.25 2.25 0 0015.75 11h-2.5z"
                fill-rule="evenodd"
              />
            </svg>
          </button>
          <button
            class="-m-2 ml-4 p-2 text-secondary hover:text-primary sm:ml-6 lg:hidden"
            type="button"
          >
            <span
              class="sr-only"
            >
              Filters
            </span>
            <svg
              aria-hidden="true"
              class="h-5 w-5"
              fill="currentColor"
              viewBox="0 0 20 20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M2.628 1.601C5.028 1.206 7.49 1 10 1s4.973.206 7.372.601a.75.75 0 01.628.74v2.288a2.25 2.25 0 01-.659 1.59l-4.682 4.683a2.25 2.25 0 00-.659 1.59v3.037c0 .684-.31 1.33-.844 1.757l-1.937 1.55A.75.75 0 018 18.25v-5.757a2.25 2.25 0 00-.659-1.591L2.659 6.22A2.25 2.25 0 012 4.629V2.34a.75.75 0 01.628-.74z"
                fill-rule="evenodd"
              />
            </svg>
          </button>
        </div>
      </div>
      <section
        aria-labelledby="products-heading"
        class="pt-6 pb-24"
      >
        <h2
          class="sr-only"
          id="products-heading"
        >
          Products
        </h2>
        <div
          class="grid grid-cols-1 gap-x-8 gap-y-10 lg:grid-cols-4"
        >
          <form
            class="hidden lg:block"
          >
            <div
              class="border-b border-gray-200 py-6"
              data-headlessui-state="open"
            >
              <h3
                class="-my-3 flow-root"
              >
                <button
                  aria-controls="headlessui-disclosure-panel-:r1:"
                  aria-expanded="true"
                  class="flex w-full items-center justify-between bg-background py-3 text-sm text-primary hover:text-secondary"
                  data-headlessui-state="open"
                  id="headlessui-disclosure-button-:r0:"
                  type="button"
                >
                  <span
                    class="font-medium text-primary"
                  >
                    Category
                  </span>
                  <span
                    class="ml-6 flex items-center"
                  >
                    <svg
                      aria-hidden="true"
                      class="h-5 w-5"
                      fill="currentColor"
                      viewBox="0 0 20 20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        clip-rule="evenodd"
                        d="M4 10a.75.75 0 01.75-.75h10.5a.75.75 0 010 1.5H4.75A.75.75 0 014 10z"
                        fill-rule="evenodd"
                      />
                    </svg>
                  </span>
                </button>
              </h3>
              <div
                class="pt-6"
                data-headlessui-state="open"
                id="headlessui-disclosure-panel-:r1:"
              >
                <div
                  class="space-y-4"
                >
                  <div
                    class="flex items-center"
                  >
                    <input
                      checked=""
                      class="h-4 w-4 rounded border-gray-300 text-primary focus:ring-secondary"
                      id="filter-category-0"
                      name="category[]"
                      type="checkbox"
                      value="men"
                    />
                    <label
                      class="ml-3 text-sm text-primary select-none"
                      for="filter-category-0"
                    >
                      Men
                    </label>
                  </div>
                  <div
                    class="flex items-center"
                  >
                    <input
                      class="h-4 w-4 rounded border-gray-300 text-primary focus:ring-secondary"
                      id="filter-category-1"
                      name="category[]"
                      type="checkbox"
                      value="women"
                    />
                    <label
                      class="ml-3 text-sm text-primary select-none"
                      for="filter-category-1"
                    >
                      Women
                    </label>
                  </div>
                  <div
                    class="flex items-center"
                  >
                    <input
                      class="h-4 w-4 rounded border-gray-300 text-primary focus:ring-secondary"
                      id="filter-category-2"
                      name="category[]"
                      type="checkbox"
                      value="kids"
                    />
                    <label
                      class="ml-3 text-sm text-primary select-none"
                      for="filter-category-2"
                    >
                      Kids
                    </label>
                  </div>
                  <div
                    class="flex items-center"
                  >
                    <input
                      class="h-4 w-4 rounded border-gray-300 text-primary focus:ring-secondary"
                      id="filter-category-3"
                      name="category[]"
                      type="checkbox"
                      value="toys"
                    />
                    <label
                      class="ml-3 text-sm text-primary select-none"
                      for="filter-category-3"
                    >
                      Toys
                    </label>
                  </div>
                  <div
                    class="flex items-center"
                  >
                    <input
                      class="h-4 w-4 rounded border-gray-300 text-primary focus:ring-secondary"
                      id="filter-category-4"
                      name="category[]"
                      type="checkbox"
                      value="electronics"
                    />
                    <label
                      class="ml-3 text-sm text-primary select-none"
                      for="filter-category-4"
                    >
                      Electronics
                    </label>
                  </div>
                  <div
                    class="flex items-center"
                  >
                    <input
                      class="h-4 w-4 rounded border-gray-300 text-primary focus:ring-secondary"
                      id="filter-category-5"
                      name="category[]"
                      type="checkbox"
                      value="home"
                    />
                    <label
                      class="ml-3 text-sm text-primary select-none"
                      for="filter-category-5"
                    >
                      Home
                    </label>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="border-b border-gray-200 py-6"
              data-headlessui-state=""
            >
              <h3
                class="-my-3 flow-root"
              >
                <button
                  aria-expanded="false"
                  class="flex w-full items-center justify-between bg-background py-3 text-primary hover:text-secondary"
                  data-headlessui-state=""
                  id="headlessui-disclosure-button-:r2:"
                  type="button"
                >
                  <span
                    class="font-medium text-primary"
                  >
                    Price
                  </span>
                  <span
                    class="ml-6 flex items-center"
                  >
                    <svg
                      aria-hidden="true"
                      class="h-5 w-5"
                      fill="currentColor"
                      viewBox="0 0 20 20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M10.75 4.75a.75.75 0 00-1.5 0v4.5h-4.5a.75.75 0 000 1.5h4.5v4.5a.75.75 0 001.5 0v-4.5h4.5a.75.75 0 000-1.5h-4.5v-4.5z"
                      />
                    </svg>
                  </span>
                </button>
              </h3>
            </div>
          </form>
          <div
            class="lg:col-span-3"
          >
            <div
              class="h-96 rounded-lg border-4 border-dashed border-gray-200 lg:h-full"
            />
          </div>
        </div>
      </section>
    </main>
  </div>
</div>
`;
